File: python/structures/src/trees/traversal.py
def _preorderHelper(node):
    output = ''

    output += str(node.value) + ' '

    if node.left != None:
        output += _preorderHelper(node.left)

    if node.right != None:
        output += _preorderHelper(node.right)

    return output

def preorder(node):
    return _preorderHelper(node).strip()

def _inorderHelper(node):
    output = ''

    if node.left != None:
        output += _inorderHelper(node.left)

    output += str(node.value) + ' '

    if node.right != None:
        output += _inorderHelper(node.right)

    return output

def inorder(node):
    return _inorderHelper(node).strip()

def _postorderHelper(node):
    output = ''

    if node.left != None:
        output += _postorderHelper(node.left)

    if node.right != None:
        output += _postorderHelper(node.right)

    output += str(node.value) + ' '

    return output

def postorder(node):
    return _postorderHelper(node).strip()

# prettier view which displays the tree as trees are commonly drawn
# TODO pretty-print logic
def treeview(node):
    print(node.value)

File: haskell/src/Tree/Traversal.hs
module Tree.Traversal where

import Tree.BST as BST

preorder :: BST.Tree a -> [a]
preorder BST.EmptyTree      = []
preorder (BST.Node x l r _) = [x] ++ preorder l ++ preorder r

inorder :: BST.Tree a -> [a]
inorder BST.EmptyTree      = []
inorder (BST.Node x l r _) = inorder l ++ [x] ++ inorder r

postorder :: BST.Tree a -> [a]
postorder BST.EmptyTree      = []
postorder (BST.Node x l r _) = postorder l ++ postorder r ++ [x]
